lib/deadline_thread_pool.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class DeadlineThreadPool
+  DEFAULT_DEADLINE = 60 # seconds
+
+  def initialize(size, job_deadline_seconds = DEFAULT_DEADLINE, max_jobs=1000)
+    @size = size
+    @job_deadline_seconds = job_deadline_seconds
+    @jobs = SizedQueue.new(max_jobs)
+    @mutex = Mutex.new
+    @condition = ConditionVariable.new
+    @running_deadlines = {}
+  end
+
+  def start
+    return false if @jobs.closed?
+
+    # Create worker threads
+    @pool = Array.new(@size) do
+      Thread.new do
+        while (job = @jobs.pop)
+          run_job(job)
+        end
+      end
+    end
+
+    # Start monitor thread
+    @monitor = Thread.new do
+      @mutex.synchronize do
+        while !(@jobs.closed? && @jobs.empty? && @running_deadlines.empty?)
+          wait_time = check_deadlines
+          if wait_time.nil?
+            @condition.wait(@mutex)
+          else
+            @condition.wait(@mutex, wait_time)
+          end
```

Comment:
`wait_time` here is not necessarily maximum amount of time the method will take.  It is the time that the condition variable will wait for the signal.  The condition variable then must reacquire the mutex before returning, and that can block for an arbitrary amount of time.  In this case, probably not long, so this should be safe.

Reply:
yeah ultimately, critical sections have to be short. Not much to be done about that, the same can be said about how queue timed pop has to  protect its data structures. This approach effectively absorbs it.